### PR TITLE
perf: get trade quotes in parallel with 10s timeout

### LINKED ISSUE #5017: Fix hyper-slow quotes
### Overview

Quotes are fetched sequentially, which means the time it takes to fetch a quote is the SUM of the durations instead of `Max(...requests)`. Additionally, if 1 quote stalls, the whole api locks up until that request fails.

Do the following:
1. use `Promise.all()` for quote fetches
2. add a 20s timeout to all quotes and insert a `SwapErrorRight` for a network timeout in the case it times out

### References and additional details

Quotes are currently fetched sequentially, which isn't ideal for performance:
image

### Acceptance Criteria

1. Quotes are fetched in parallel
2. The maximum time take to fetch all quotes is 20s total
3. Timeouts are added to the result in the form of `SwapErrorRight` monad

### Need By Date

_No response_

### Screenshots/Mockups

_No response_

### Estimated effort

3h

### PR DESCRIPTION
**Pull request by woodenfurniture**
## Description

Fetches trade quotes in parallel and times out after 10s such that 1 slow service does not lock up the whole trade quote flow.

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #5017

## Risk

Low risk of quotes broken

## Testing

Check quotes are fetched.

### Engineering

See above

### Operations

See above

## Screenshots (if applicable)

Able to get quote
image

### CODE REVIEW

**Comment by woodenfurniture** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #5015** <a href="https://app.graphite.dev/github/pr/shapeshift/web/5015" target="_blank">Graphite</a> 
    * **PR #5016** <a href="https://app.graphite.dev/github/pr/shapeshift/web/5016" target="_blank">Graphite</a> 
      * **PR #5018** <a href="https://app.graphite.dev/github/pr/shapeshift/web/5018" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/5018?utm_source=stack-comment).

File: src/state/apis/swappers/swappersApi.ts [hunk 1]
```diff
@@ -18,6 +19,26 @@ import { getOneInchTradeQuoteHelper } from './helpers/getOneInchTradeQuoteHelper
 import { getOsmosisTradeQuoteHelper } from './helpers/getOsmosisTradeQuoteApiHelper'
 import { getZrxTradeQuoteHelper } from './helpers/getZrxTradeQuoteHelper'
 
+const QUOTE_TIMEOUT_MS = 10_000
+
+const TIMEOUT_ERROR = makeSwapErrorRight({
+  message: `quote timed out after ${QUOTE_TIMEOUT_MS / 1000}s`,
+})
+
+const timeout = <Left, Right>(
```
Comment (0xdef1cafe) [inline src/state/apis/swappers/swappersApi.ts:28]: where have i seen THIS before

Comment (gomesalexandre) [reply src/state/apis/swappers/swappersApi.ts:28]: beautiful

File: src/state/apis/swappers/swappersApi.ts [hunk 2]
```diff
@@ -30,83 +51,57 @@ export const swappersApi = createApi({
...
+        const quotes = await Promise.all(
+          swappers
+            .filter(({ isEnabled }) => isEnabled)
+            .map(({ swapperName, getTradeQuote }) =>
+              timeout<TradeQuote2, SwapErrorRight>(
+                getTradeQuote(getTradeQuoteInput, state),
+                TIMEOUT_ERROR,
+              ).then(quote => ({
+                ...quote,
+                swapperName,
```
Comment (gomesalexandre) [inline src/state/apis/swappers/swappersApi.ts:97]: Definitely not for this PR but this broke my brain, I was expecting this to be an akschual quote not a result, if we want to make things cleaner in the future we could give a better home to the quote in a `{quoteResult: quote, swapperName}` resolution, but very very nitpicky, spreading the result object and adding a property to it is perfectly fine

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
Tested locally with app env, as well as cross-tested against develop build and remote release:

### Confirmed *most* quotes are loading and are doing so blazingly fast:

image
image
image

### Noticed xDAI to any asset (same/cross-chain) is showing regressions against prod

- Prod

image
image

- This diff

image
image

Same thing in the other direction (xDAI / ERC-20 to any Gnosis chain / cross-chain asset) which seems to indicate Gnosis is borked altogether as a swapping chain either from this PR or https://github.com/shapeshift/web/pull/5016/commits

### Osmosis quotes are borked in both directions

- Develop/Prod

image
image


- This diff

image

### ETH on Ethereum -> ETH on Optimism is broken

- This diff

image

- Develop/Prod

image

### FOX -> WETH gives no quotes

- This diff

image

- Prod

image

Retested all above against https://github.com/shapeshift/web/pull/5016 and confirmed these are actual regressions, as I couldn't repro in 5016, requesting changes so we don't get this in for now

**Review by gomesalexandre — APPROVED** [review]
The issue noted in the previous CR has been fixed in 2bc4a877393276e3cf5e60c4d3958f77121ff7f1

The reason why this was failing is we were assuming all failures to be handled by the error monad, while some errors are actually rejected promises, not :err tuple e.g

```typescript
swappersApi.ts:108 SwapError: [getPriceRatio]: No buyPoolId found for asset eip155:1/erc20:0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2
```

Will make `getPriceRatio` be a proper Result monad without any promise rejections in a follow-up, but for the time being, this is the most atomic approach to implement this

- ATOM - OSMO

image

- OSMO -> ATOM

image

- FOX -> WETH

image

- ETH -> Optimism ETH

image

**Pull request merged**